contrib/platform/src/com/sun/jna/platform/mac/SystemB.java

```diff
@@ -742,28 +815,53 @@ int sysctlbyname(String name, Pointer oldp, IntByReference oldlenp,
      *            Pointer to number of elements in the returned structure
      * @return 0 on success; sets errno on failure
      */
-    int host_processor_info(int machPort, int flavor, IntByReference procCount,
-        PointerByReference procInfo, IntByReference procInfoCount);
+    int host_processor_info_ptr(MachPort machPort, int flavor, IntByReference procCount, PointerByReference procInfo,
```

Comment:
I had a look at the documentation and host_processor_info does not take a mach_port_t but a host_t. I don't see the relationship between the two (https://developer.apple.com/documentation/kernel/1502854-host_processor_info?language=objc https://developer.apple.com/documentation/kernel/host_t?language=objc)

Binding this as a Pointer/PointerType is the right way none the less, as the int is definitely broken.

Reply:
Hosts, tasks, and mach ports are all generally the same type.... pointer-sized unsigned ints defined as pointers to empty structures, which translate to full structures on the kernel side.  Agree this might not be "mach port" type, and perhaps there's room for host and task types as well, but I think `NativeLong` may be preferred to `PointerType` because they don't really point to anything.

Reply:
Be careful! The gstreamer bindings were broken, because someone made the same assumptions and that broke hard. The core datastructure was read with a NativeLong, but was a Pointer. This was noticed, until it exploded on windows where sizeof(void*) != sizeof(long)

Interesting read about the madness: http://nickdesaulniers.github.io/blog/2016/05/30/data-models-and-word-size/

These bugs are subtle and have the tendency to lead to chases into the wrong direction.

Reply:
Fascinating story.  OK, I'll stick with the pointers.

Reply:
But then there's this.  `natural_t` is the type of the second argument to `mach_port_deallocate()`, which as I've shown earlier takes `io_object_t`.
```
/*
 * A natural_t is the type for the native
 * integer type, e.g. 32 or 64 or.. whatever
 * register size the machine has.  Unsigned, it is
 * used for entities that might be either
 * unsigned integers or pointers, and for
 * type-casting between the two.
 * For instance, the IPC system represents
 * a port in user space as an integer and
 * in kernel space as a pointer.
 */
typedef unsigned int	natural_t;
```

Reply:
From: http://web.mit.edu/darwin/src/modules/xnu/osfmk/man/mach_port_deallocate.html

the second argument is _not_ a `mach_port_t`, but a `mach_port_name_t`.  With the above definition `mach_port_name_t` is a 32bit wide integer (I have not yet seen a different sized integer. This could be a java `int` or a custom type based on `IntegerType`.

Reply:
Ah, yes... I see where I saw "native integer" and thought native long.  This does eventually track back to `int`.  And the other types trace back to pointers.  I think I need more coffee.

Reply:
I still can't understand how this code works since the second argument of mach_port_deallocate should be an int, not a pointer:
```
kern_return_t
IOObjectRelease(
	io_object_t	object )
{
    return( mach_port_deallocate( mach_task_self(), object ));
}
```
But I'll just pretend I didn't see it. :)

Reply:
And earlier you linked to [this source](https://www.gnu.org/software/hurd/gnumach-doc/Port-Destruction.html) which has: 
```
kern_return_t mach_port_deallocate (ipc_space_t task, mach_port_t name)
T
```
But GNU is not XNU so perhaps that's a BSD-ism and not macOS...

And if the argument is an `int` then it does not tell me how to translate the actual `mach_port_t` argument (if it's modeled as a pointer).

Still digging.

Reply:
> I had a look at the documentation and host_processor_info does not take a mach_port_t but a host_t. I don't see the relationship between the two

That probably came from [here](https://fergofrog.com/code/cbowser/xnu/osfmk/mach/mach_types.h.html)
```
typedef mach_port_t		task_t;
typedef mach_port_t		thread_t;
typedef mach_port_t		host_t;
```
Although that doesn't match the header files on my machine:
```
typedef struct task                     *task_t, *task_name_t, *task_inspect_t, 
typedef struct thread           *thread_t, *thread_act_t, *thread_inspect_t;
typedef struct host                     *host_t;
```
Where the structures are empty:
```
struct task;
struct thread;
struct host;
```

Reply:
> I had a look at the documentation and host_processor_info does not take a mach_port_t but a host_t. I don't see the relationship between the two

See the source for [mach_types.h](https://fergofrog.com/code/cbowser/xnu/osfmk/mach/mach_types.h.html) which has two different refs, one for inside the kernel and one for user-space.

For the kernel, these are pointers:
```
/*
* If we are in the kernel, then pick up the kernel definitions for
* the basic mach types.
*/
typedef struct task			*task_t, *task_name_t, *task_inspect_t, *task_suspension_token_t;
typedef struct thread		*thread_t, *thread_act_t, *thread_inspect_t;
typedef struct host			*host_t;
```
But in userspace:
```
/*
* If we are not in the kernel, then these will all be represented by
* ports at user-space.
*/
typedef mach_port_t		task_t;
typedef mach_port_t		thread_t;
typedef mach_port_t		host_t;
```

Reply:
OK, I'm pretty sure I've got this all (mostly) sorted out.

Outside the kernel, all the various port-related types map to `mach_port_t` (which itself is an `ipc_port_t` which is a pointer to an empty `ipc_port` structure).  So the mapping of `MachPort` to `PointerType` is appropriate.  We could define `MachTask`, `MachThread`, and `MachHost` types as subclasses.

The `mach_port_name_t` is a (unsigned) 32-bit integer and is derived from the `mach_port_t` pointer by typecasting:
`#define CAST_MACH_PORT_TO_NAME(x) ((mach_port_name_t)(uintptr_t)(x))`

Reply:
OK, I've committed based on the above and think it's the correct user-space interpretation.  Ready for another review and/or lesson. :)